herd/AArch64Arch_herd.ml

```diff
@@ -373,6 +373,32 @@ module Make (C:Arch_herd.Config)(V:Value.AArch64) =
           module FaultType = FaultType.AArch64
         end)
 
+    let convert_if_imm_branch pc l2i same_proc instr =
+      let open BranchTarget in
+      let labelmap l =
+        let tgt =
+          try Label.Map.find l l2i 
+          with Not_found -> Warn.user_error
+            "Label %s not found, although used in a branch %s"
+            l (pp_instruction PPMode.Ascii instr)
+          in
+        if same_proc pc tgt then
+          tgt - pc
+        else
+          Warn.user_error
+            "The branch %s and its destination label %s must be on the same proc"
+            (pp_instruction PPMode.Ascii instr) l
+        in
+      match instr with
+      | I_B (Lbl l) -> I_B (Offset (labelmap l))
+      | I_BC (c,(Lbl l)) -> I_BC (c,(Offset (labelmap l)))
+      | I_BL (Lbl l) -> I_BL (Offset (labelmap l))
+      | I_CBNZ (v,r,(Lbl l)) -> I_CBNZ (v,r,(Offset (labelmap l)))
+      | I_CBZ (v,r,(Lbl l)) -> I_CBZ (v,r,(Offset (labelmap l)))
+      | I_TBNZ (v,r,k,(Lbl l)) -> I_TBNZ (v,r,k,(Offset (labelmap l)))
+      | I_TBZ (v,r,k,(Lbl l)) -> I_TBZ (v,r,k,(Offset (labelmap l)))
```

Comment:
That's a good catch -- thank you, @maranget!

For a minute I was not sure if this conversion is strictly necessary, so I am adding a new test `AArch64/A249` to record that it is. (By analogy with the test in the PR description.)

Reply:
Note that several  AArch64 tests fail after applying conversion to ADR: A139, L023, L050, L051. Out of these, L023 needs a trivial fix (that is, removing ADR). More detail below.

Firstly, the case of `A139`. Notice that the test uses a pointer instead of a label. I am not sure I understand the intent here, I am afraid... Is the idea that pointers are PC-relative addresses? In that case, should herd treat `pointer_to_x` as a label then (and not convert it, accordingly)?
```
AArch64 A139
(* Tests whether we can load from a pointer - symbolic location+offset *)

{ int *pointer_to_x = x; int x[2]={4,42} }

P0;
  ADR X1, pointer_to_x; (* loads pointer_to_x into X1 *)
  LDR X1, [X1];         (* loads x into X1 *)
  LDR W0, [X1, #4];     (* loads x[1] into X1 *)

exists (0:X0 = 42)
```

Secondly, the case of L050 and L051. They now fail and illustrate several points. Consider the following test:
```
AArch64 L050-truncated
{
0:X0=instr:"B.EQ .+8";
}
  P0         ;
L0:          ;
 B .+4       ;
 ADR X1,L0   ;
locations [0:X1]
```
Currently herd outputs a herd-internal numerical address here (`0:X1=10000`), which leaks an implementation detail to end-user. Furthermore, it creates problems if followed up by a memory access: loads end up being unmatched, which is what happens in `L050` and `L051`, and stores complain at a store to a non-symbolic location.

Should ADR be able to return a non-label? An approach I want to try is to make ADR return a label always, so that the result of `ADR X1,L0` is then equivalent to having `0:X1=P0:L0` in precondition.